sdk/provisioning/Azure.Provisioning.CloudMachine/src/CloudMachineInfrastructure.cs

```diff
@@ -0,0 +1,216 @@
+                ],
+                IsAdvancedFilteringOnArraysEnabled = true
+            },
+            EventDeliverySchema = EventDeliverySchema.EventGridSchema,
+            RetryPolicy = new EventSubscriptionRetryPolicy
+            {
+                MaxDeliveryAttempts = 30,
+                EventTimeToLiveInMinutes = 1440
+            }
+        };
+    }
+
+    public override ProvisioningPlan Build(ProvisioningContext? context = null)
+    {
+        // Always add a default location parameter.
+        // azd assumes there will be a location parameter for every module.
+        // The Infrastructure location resolver will resolve unset Location properties to this parameter.
+        Add(new BicepParameter("location", typeof(string))
+        {
+            Description = "The location for the resource(s) to be deployed.",
+            Value = BicepFunction.GetResourceGroup().Location
+        });
+
+        Add(PrincipalIdParameter);
+        Add(PrincipalTypeParameter);
+        Add(PrincipalNameParameter);
+
+        Add(_identity);
+        Add(_storage);
+        Add(_storage.AssignRole(StorageBuiltInRole.StorageBlobDataContributor, RoleManagementPrincipalType.User, PrincipalIdParameter));
+        Add(_storage.AssignRole(StorageBuiltInRole.StorageTableDataContributor, RoleManagementPrincipalType.User, PrincipalIdParameter));
+        Add(_container);
+        Add(_blobs);
+        Add(_serviceBusNamespace);
+        Add(_serviceBusNamespace.AssignRole(ServiceBusBuiltInRole.AzureServiceBusDataOwner, RoleManagementPrincipalType.User, PrincipalIdParameter));
+        Add(_serviceBusNamespaceAuthorizationRule);
+        Add(_serviceBusTopic_main);
+        Add(_serviceBusTopic_app);
+        Add(_serviceBusSubscription_main);
+        Add(_serviceBusSubscription_app);
+
+        // This is necessary until SystemTopic adds an AssignRole method.
+        var role = ServiceBusBuiltInRole.AzureServiceBusDataOwner;
+        RoleAssignment roleAssignment = new RoleAssignment(_eventGridTopic_Blobs.ResourceName + "_" + _identity.ResourceName + "_" + ServiceBusBuiltInRole.GetBuiltInRoleName(role));
+        roleAssignment.Name = BicepFunction.CreateGuid(_eventGridTopic_Blobs.Id, _identity.Id, BicepFunction.GetSubscriptionResourceId("Microsoft.Authorization/roleDefinitions", role.ToString()));
+        roleAssignment.Scope = new IdentifierExpression(_eventGridTopic_Blobs.ResourceName);
+        roleAssignment.PrincipalType = RoleManagementPrincipalType.ServicePrincipal;
+        roleAssignment.RoleDefinitionId = BicepFunction.GetSubscriptionResourceId("Microsoft.Authorization/roleDefinitions", role.ToString());
+        roleAssignment.PrincipalId = _identity.PrincipalId;
+        Add(roleAssignment);
+        Add(_systemTopicEventSubscription);
+        Add(_eventGridTopic_Blobs);
+
+        // Placeholders for now.
+        Add(new BicepOutput($"storage_name", typeof(string)) { Value = _storage.Name });
```

Comment:
Do you want the `name` or the primary endpoint like `_storage.PrimaryEndpoints.Value!.BlobUri`?

Reply:
just the name, since the endpoint can be derrived.

Reply:
Not if you want this to work on more than public Azure in the future.  :smile: